### PR TITLE
Hw1

### PR DESCRIPTION
**Pull request by dekatzenel**

### CODE REVIEW

File: HW1/P2/P2.txt [hunk 1]
```diff
@@ -1 +1,9 @@
...
+
+After calling partitionBy(), all partitions took between 2,020,000 and 2,045,000 iterations. This means that the worst partitions took less than 1.0124x the iterations of the best partitions. It is fair to say, therefore, that all partitions took roughly the same number of iterations.
+
+However, without calling partitionBy(), the work prior to drawing took 1.6 min total, almost all of it spent in executor computing time.
+
+On the other hand, with calling partitionBy(), the work prior to drawing can be divided into two parts. The partitionBy() call, which takes 4s, is almost entirely composed of executor computing time (4s) but includes 13.5KB of shuffle write. The rest of the computation takes 1.8 min, almost all executor computing time, and includes 947.5KB of shuffle read. Therefore, the better-parallelized version is about 0.2 to 0.3 minutes longer than the default. This time could be made up if multiple computations were run on the shuffled data.
```
Comment (drishi) [inline HW1/P2/P2.txt]: Good analysis here, interesting to see the exact # of iterations and the percentage breakdown!

File: HW1/P2/P2b.py [hunk 2]
```diff
@@ -1,3 +1,23 @@
 from P2 import *
```
Comment (drishi) [inline HW1/P2/P2b.py]: My histogram looked the exact same.

File: HW1/P3/P3.txt [hunk 3]
```diff
@@ -1 +1 @@
-# Your discussion here
+[((u'a', u'e', u'r', u's', u't'), 11, [u'arets', u'aster', u'rates', u'reast', u'resat', u'stare', u'stear', u'strae', u'tares', u'tears', u'teras'])]
```
Comment (drishi) [inline HW1/P3/P3.txt]: Nice, I keyed by word rather than separate chars but that's the only difference.

File: HW1/P4/P4.txt [hunk 4]
```diff
@@ -1 +1,8 @@
...
+I use the bfs function arguments indicated, but I retrieve the context from the graph inside of the bfs function to create both an accumulator and an empty RDD.
+
+"Recall that, for a node v in a graph G, the radius r(v) is the distance between v and a reachable node farthest away from v. The diameter d(G) of a graph G is the maximum radius of nodes v ∈ G. That is, d(G) = max_v r(v)" - http://www.math.cmu.edu/~ctsourak/tkdd10.pdf
+If the diameter of a graph is 10, then the maximum distance between any two connected nodes on the graph is 10. If we consider a step to be all operations performed at distance d from the starting point, which must all be contiguous from the structure of the BFS algorithm, then the search is executed at most 10 steps.
+
+Including the origin, the number of nodes touched from each start is: 'MISS THING/MARY': 7, 'ORWELL': 9, 'CAPTAIN AMERICA': 6408
+If a character does not have a defined distance from a given origin, then that character is not reachable from that origin. This means that there is no chain of mutual appearances in a comic book that connects the two characters - their universes are disjoint.
```
Comment (drishi) [inline HW1/P4/P4.txt]: Yep exactly. Read someone else's response where it said that this graph is completely connected and so this isn't actually possible on this specific graph, which was interesting.

File: HW1/P5/P5.txt [hunk 5]
```diff
@@ -1 +1,4 @@
-# Your discussion here
+One of the shortest paths from Kevin Bacon to Harvard University is: Kevin_Bacon -> College_Bowl -> Harvard_University, where a -> b means "a links to b".
+One of the shortest paths from Harvard University to Kevin Bacon is: Harvard_University -> 1998 -> 1958 -> Kevin_Bacon
+
+My code for connected components was not fast enough to arrive at an answer
```
Comment (drishi) [inline HW1/P5/P5.txt]: I actually got a different set of answers here. Shortest path from Kevin Bacon to Harvard for me was several steps long (College Bowl went to John Lithgow to Marisa Silver to 6 degrees of separation to Time), and for Harvard to Kevin there were 120 paths.

File: HW1/P5/P5_connected_components.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+            #Update record of previously visited nodes
+            assert copartitioned(graph, toVisit)
+            graph = graph.subtract(toVisit).partitionBy(16).cache()
+            visited = visited.union(toVisit).partitionBy(16)
+            #Update record of nodes to visit
+            neighbors = toVisit.values().flatMap(lambda x: x).collect()
+            toVisit = graph.filter(lambda KV: KV[0] in neighbors).cache()
+            assert copartitioned(graph, toVisit)
+        result.append(visited.count())
+    return result
```
Comment (drishi) [inline HW1/P5/P5_connected_components.py]: Wow this is concise CC code! Mine wasn't fast enough to run either :(

File: HW1/P6/P6.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+for start in starts:
+    current_tuple = start
+    shakespeare_sentence = current_tuple[0][0] + ' ' + current_tuple[0][1]
+    for j in range(2, 20):
+        next_word = weighted_choice(current_tuple[1])
+        current_tuple_precursor = markov_chain.filter(lambda KV: KV[0] == (current_tuple[0][1], next_word))
+        current_tuple = current_tuple_precursor.first()
+        shakespeare_sentence = shakespeare_sentence + ' ' + next_word
+    sentences.append(shakespeare_sentence)
+print sentences
```
Comment (drishi) [inline HW1/P6/P6.py]: Nice! Well done.

File: HW1/P2/P2.txt [hunk 8]
```diff
@@ -1 +1,9 @@
-# Your discussion here
+To divide the workload more evenly, I applied the default partitioning hash of .partitionBy() to each of the x and y axes prior to calculating their cartesian product. (I also tried several variations on log, exp, and pow as partitioning functions, but all of them were worse at dividing the work evenly than not calling .partitionBy() at all, so I used the simpler strategy.)
+
+Without calling partitionBy(), 85-90% of the partitions took less than 5,000,000 iterations total, but the remaining partitions took up to about 20,000,000 total. This means that the worst partitions took at least 4x the iterations of a typical partition, possibly significantly more because the histogram makes it impossible to distinguish between, e.g., 1,000,000 and 4,000,000 iterations.
+
+After calling partitionBy(), all partitions took between 2,020,000 and 2,045,000 iterations. This means that the worst partitions took less than 1.0124x the iterations of the best partitions. It is fair to say, therefore, that all partitions took roughly the same number of iterations.
```
Comment (zelong430) [inline HW1/P2/P2.txt]: Very detailed analysis

File: HW1/P3/P3.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
+from pyspark import SparkContext, SparkConf
+
+#Setup
+conf = SparkConf().setAppName("anagrams").setMaster("local[*]")
+sc = SparkContext(conf=conf)
+
+#For each word, create a key out of its ordered sequence of letters
+words = sc.textFile('EOWL_words.txt')
+sortedLetters_words = words.map(lambda w: (tuple(sorted(w)), [w]))
```
Comment (zelong430) [inline HW1/P3/P3.py]: Nice implementation of tuple. There is a performance gain by using tuple instead of list.

File: HW1/P4/P4.txt [hunk 10]
```diff
@@ -1 +1,8 @@
-# Your discussion here
+My graph is represented as an RDD of key-value pairs. The key is the name of the comic book character as found in the input file and the value is a set of the names of all other comic book characters who share at least one issue with that character, as the names appear in the input file.
+I use the bfs function arguments indicated, but I retrieve the context from the graph inside of the bfs function to create both an accumulator and an empty RDD.
+
+"Recall that, for a node v in a graph G, the radius r(v) is the distance between v and a reachable node farthest away from v. The diameter d(G) of a graph G is the maximum radius of nodes v ∈ G. That is, d(G) = max_v r(v)" - http://www.math.cmu.edu/~ctsourak/tkdd10.pdf
```
Comment (zelong430) [inline HW1/P4/P4.txt]: Nice and comprehensive discussion here.

File: HW1/P5/P5.txt [hunk 11]
```diff
@@ -1 +1,4 @@
-# Your discussion here
+One of the shortest paths from Kevin Bacon to Harvard University is: Kevin_Bacon -> College_Bowl -> Harvard_University, where a -> b means "a links to b".
```
Comment (zelong430) [inline HW1/P5/P5.txt]: Different path generated. But I'm sure that there exist a lot of 3-step paths.

File: HW1/P6/P6.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+#Taken from http://stackoverflow.com/questions/3679694/a-weighted-version-of-random-choice
+def weighted_choice(choices):
+    total = sum(w for c, w in choices)
+    r = random.uniform(0, total)
+    upto = 0
+    for c, w in choices:
+          if upto + w >= r:
+             return c
+          upto += w
+    assert False, "Shouldn't get here"
```
Comment (adisuissa) [inline HW1/P6/P6.py:34]: Nice work.
You can use numpy's random.choice function to clarify the weighted_choice function.

**Pull request closed without merging**